### PR TITLE
print the first assertion failure of each test, instead of the last. Fixes #121, #220

### LINKED ISSUE #121: Output every failed assertion or just the first one (currently outputs just the last one)
```js
test('foo', t => {
	t.plan(2);
	var a = 2;
	var b = 3;
	var c = 4;
	t.ok(a === b);
	t.ok(b === c);
});
```

output:
```
 ✖ foo   
  t.ok(b === c)
       |     | 
       3     4 
  

  1 test failed
```

It would be better if either:
 1. Both assertion failures printed.
 2. If there can only be one, print the first.

### LINKED ISSUE #220: Stop on first failed assertion
I noticed that test cases are not stopped when an assertion fails. To be exact, they are stopped asynchronously. Here is an example:

```
test(t => {
  let arr = [ { name: 'abc' } ];
  t.is(arr.length, 2);
  t.is(arr[1].name, 'xyz');
  t.end();
});
```

The error I'd expect to see is "1 !== 2" while the actual error I get is "cannot read property 'name' of undefined" because execution is not stopped on the failed assertion (on the line above).

Is there a reason why `Test#exit()` is called asynchronously (by `Test#_assert()`) rather than synchronously (which is the case for errors thrown outside assertions)?

### PR DESCRIPTION
**Pull request by jamestalmage**
Fixes #121.
Fixes #220.

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -312,3 +312,33 @@ test('wait for test to end', function (t) {
 		avaTest.pass();
 	}, 1234);
 });
+
+test('promise is rejected with the first assertError', function (t) {
```
Comment (sindresorhus) [inline test/test.js]: Not sure why `promise` is mentioned here. It's not really relevant. Should be more like:

> fails with the first assertError

Sem with the below.

File: lib/test.js [hunk 2]
```diff
@@ -68,20 +69,29 @@ Object.keys(assert).forEach(function (el) {
...
 	};
 });
 
+Test.prototype._setAssertError = function (err) {
+	if (this.assertError === undefined) {
+		if (err === undefined) {
+			err = 'undefined';
+		}
+		this.assertError = err;
+	}
```
Comment (sindresorhus) [inline lib/test.js]: I usually prefer early returns over nesting when possible:

```js
if (this.assertError !== undefined) {
	return;
}

if (err === undefined) {
	err = 'undefined';
}

this.assertError = err;
```

**Comment by sindresorhus** [comment]
@jamestalmage Looks good except for some minor nitpick :) Would you mind adding the assertion failure behavior to the docs so it's explicit how it works?

When you reference tickets in the the title, can you do so in the description too? It makes it easier to click through. The issue references in the title aren't clickable.

@vdemedes ?

**Comment by jamestalmage** [comment]
> When you reference tickets in the the title, can you do so in the description too? It makes it easier to click through. The issue references in the title aren't clickable.

Errr...

> James Talmage commented 6 hours ago ...

**Comment by sindresorhus** [comment]
> Errr...

Ah yes. I should have mentioned. I added the issue references to the PR description for you here.

**Comment by jamestalmage** [comment]
OK, I just followed up on a number of other PR's

**Comment by sindresorhus** [comment]
Can you fix the merge conflict?

**Comment by jamestalmage** [comment]
Working on that and the documentation.

File: test/test.js [hunk 3]
```diff
@@ -312,3 +312,33 @@ test('wait for test to end', function (t) {
...
+		a.is(1, 2);
+		a.is(3, 4);
+	}).run().catch(function (err) {
+		t.is(err.actual, 1);
+		t.is(err.expected, 2);
+		t.end();
+	});
+});
+
+test('fails with thrown falsie value', function (t) {
```
Comment (sindresorhus) [inline test/test.js]: super minor, but for consistency `falsie` → `falsy`

**Comment by sindresorhus** [comment]
@jamestalmage You are so productive! :D

**Comment by jamestalmage** [comment]
OK, I think that covers it.

**Pull request closed without merging**